From 0015b7b66cba9351f0413c75ab924b3a2ccc40e2 Mon Sep 17 00:00:00 2001
From: Mohit Pant <mohipant@qti.qualcomm.com>
Date: Wed, 4 Feb 2026 00:34:05 -0800
Subject: [PATCH 12/40] qcacld-3.0: Deep copy assoc req to avoid race condition

During association, a race condition occurs when the assoc indication
for OWE update and the assoc response are processed in parallel. After
sending the assoc response, the thread frees the parsed assoc request
stored in the pe_session entry. This leaves a dangling pointer in the
assoc indication, which is still being used by another thread when
sending the indication to the upper layer.

To prevent this use-after-free condition, this change creates a deep
copy of the parsed assoc request. The copied frame is then freed safely
just before freeing the assoc indication structure.

CRs-Fixed: 4412866
Change-Id: Ibe97e0785ab8787adca2a01c5dfa907ca3fb9d95
---
 core/sap/src/sap_fsm.c          |  9 +++++++++
 core/sap/src/sap_module.c       |  8 ++++++++
 core/sme/src/csr/csr_api_roam.c | 34 +++++++++++++++++++++++++++++++--
 3 files changed, 49 insertions(+), 2 deletions(-)

diff --git a/core/sap/src/sap_fsm.c b/core/sap/src/sap_fsm.c
index 99ee9eeef3..a7d72ce05d 100644
--- a/core/sap/src/sap_fsm.c
+++ b/core/sap/src/sap_fsm.c
@@ -2612,6 +2612,9 @@ QDF_STATUS sap_signal_hdd_event(struct sap_context *sap_ctx,
 					 csr_roaminfo->owe_pending_assoc_ind,
 					 csr_roaminfo->fReassocReq)) {
 				sap_err("Failed to fill OWE IE");
+				if (csr_roaminfo->owe_pending_assoc_ind->assocReqPtr)
+					qdf_mem_free(
+						csr_roaminfo->owe_pending_assoc_ind->assocReqPtr);
 				qdf_mem_free(csr_roaminfo->
 					     owe_pending_assoc_ind);
 				csr_roaminfo->owe_pending_assoc_ind = NULL;
@@ -2621,6 +2624,9 @@ QDF_STATUS sap_signal_hdd_event(struct sap_context *sap_ctx,
 			if (!sap_save_owe_pending_assoc_ind(sap_ctx,
 					 csr_roaminfo->owe_pending_assoc_ind)) {
 				sap_err("Failed to save assoc ind");
+				if (csr_roaminfo->owe_pending_assoc_ind->assocReqPtr)
+					qdf_mem_free(
+						csr_roaminfo->owe_pending_assoc_ind->assocReqPtr);
 				qdf_mem_free(csr_roaminfo->
 					     owe_pending_assoc_ind);
 				csr_roaminfo->owe_pending_assoc_ind = NULL;
@@ -2634,6 +2640,9 @@ QDF_STATUS sap_signal_hdd_event(struct sap_context *sap_ctx,
 			if (!sap_save_ft_pending_assoc_ind(sap_ctx,
 			    csr_roaminfo->ft_pending_assoc_ind)) {
 				sap_err("Failed to save ft assoc ind");
+				if (csr_roaminfo->ft_pending_assoc_ind->assocReqPtr)
+					qdf_mem_free(
+						csr_roaminfo->ft_pending_assoc_ind->assocReqPtr);
 				qdf_mem_free(csr_roaminfo->ft_pending_assoc_ind);
 				csr_roaminfo->ft_pending_assoc_ind = NULL;
 				qdf_mem_free(sap_ap_event);
diff --git a/core/sap/src/sap_module.c b/core/sap/src/sap_module.c
index 51671f9943..6e0e9026a7 100644
--- a/core/sap/src/sap_module.c
+++ b/core/sap/src/sap_module.c
@@ -291,6 +291,8 @@ static void wlansap_owe_cleanup(struct sap_context *sap_ctx)
 			assoc_ind->owe_ie_len = 0;
 			assoc_ind->owe_status = STATUS_UNSPECIFIED_FAILURE;
 			status = sme_update_owe_info(mac, assoc_ind);
+			if (assoc_ind->assocReqPtr)
+				qdf_mem_free(assoc_ind->assocReqPtr);
 			qdf_mem_free(assoc_ind);
 		} else {
 			sap_err("Failed to remove assoc ind");
@@ -339,6 +341,8 @@ static void wlansap_ft_cleanup(struct sap_context *sap_ctx)
 			assoc_ind->ft_ie = NULL;
 			assoc_ind->ft_ie_len = 0;
 			assoc_ind->ft_status = STATUS_UNSPECIFIED_FAILURE;
+			if (assoc_ind->assocReqPtr)
+				qdf_mem_free(assoc_ind->assocReqPtr);
 			qdf_mem_free(assoc_ind);
 		} else {
 			sap_err("Failed to remove assoc ind");
@@ -3292,6 +3296,8 @@ QDF_STATUS wlansap_update_owe_info(struct sap_context *sap_ctx,
 		assoc_ind->owe_ie_len = ie_len;
 		assoc_ind->owe_status = owe_status;
 		status = sme_update_owe_info(mac, assoc_ind);
+		if (assoc_ind->assocReqPtr)
+			qdf_mem_free(assoc_ind->assocReqPtr);
 		qdf_mem_free(assoc_ind);
 	} else {
 		/*
@@ -3367,6 +3373,8 @@ QDF_STATUS wlansap_update_ft_info(struct sap_context *sap_ctx,
 		assoc_ind->ft_ie_len = ie_len;
 		assoc_ind->ft_status = ft_status;
 		status = sme_update_ft_info(mac, assoc_ind);
+		if (assoc_ind->assocReqPtr)
+			qdf_mem_free(assoc_ind->assocReqPtr);
 		qdf_mem_free(assoc_ind);
 	}
 	return status;
diff --git a/core/sme/src/csr/csr_api_roam.c b/core/sme/src/csr/csr_api_roam.c
index 34a5848aa0..7eed4706ea 100644
--- a/core/sme/src/csr/csr_api_roam.c
+++ b/core/sme/src/csr/csr_api_roam.c
@@ -4068,15 +4068,41 @@ csr_roam_chk_lnk_assoc_ind(struct mac_context *mac_ctx, tSirSmeRsp *msg_ptr)
 		if (csr_akm_type == eCSR_AUTH_TYPE_OWE) {
 			roam_info->owe_pending_assoc_ind = qdf_mem_malloc(
 							    sizeof(*pAssocInd));
-			if (roam_info->owe_pending_assoc_ind)
+			if (roam_info->owe_pending_assoc_ind) {
 				qdf_mem_copy(roam_info->owe_pending_assoc_ind,
 					     pAssocInd, sizeof(*pAssocInd));
+				if (pAssocInd->assocReqLength) {
+					roam_info->owe_pending_assoc_ind->assocReqPtr =
+						qdf_mem_malloc(pAssocInd->assocReqLength);
+					if (roam_info->owe_pending_assoc_ind->assocReqPtr) {
+						qdf_mem_copy(roam_info->owe_pending_assoc_ind->assocReqPtr,
+							     pAssocInd->assocReqPtr,
+							     pAssocInd->assocReqLength);
+					} else {
+						sme_err("OWE assocReqPtr alloc failed");
+						roam_info->owe_pending_assoc_ind->assocReqPtr = NULL;
+					}
+				}
+			}
 		} else if (csr_akm_type == eCSR_AUTH_TYPE_FT_RSN_PSK) {
 			roam_info->ft_pending_assoc_ind = qdf_mem_malloc(
 			    sizeof(*pAssocInd));
-			if (roam_info->ft_pending_assoc_ind)
+			if (roam_info->ft_pending_assoc_ind) {
 				qdf_mem_copy(roam_info->ft_pending_assoc_ind,
 					     pAssocInd, sizeof(*pAssocInd));
+				if (roam_info->ft_pending_assoc_ind->assocReqLength) {
+					roam_info->ft_pending_assoc_ind->assocReqPtr =
+						qdf_mem_malloc(pAssocInd->assocReqLength);
+					if (roam_info->ft_pending_assoc_ind->assocReqPtr) {
+						qdf_mem_copy(roam_info->ft_pending_assoc_ind->assocReqPtr,
+							     pAssocInd->assocReqPtr,
+							     pAssocInd->assocReqLength);
+					} else {
+						sme_err("FT assocReqPtr alloc failed");
+						roam_info->ft_pending_assoc_ind->assocReqPtr = NULL;
+					}
+				}
+			}
 		}
 		status = csr_roam_call_callback(mac_ctx, sessionId,
 					roam_info, eCSR_ROAM_INFRA_IND,
@@ -4084,9 +4110,13 @@ csr_roam_chk_lnk_assoc_ind(struct mac_context *mac_ctx, tSirSmeRsp *msg_ptr)
 		if (!QDF_IS_STATUS_SUCCESS(status)) {
 			/* Refused due to Mac filtering */
 			if (roam_info->owe_pending_assoc_ind) {
+				if (roam_info->owe_pending_assoc_ind->assocReqPtr)
+					qdf_mem_free(roam_info->owe_pending_assoc_ind->assocReqPtr);
 				qdf_mem_free(roam_info->owe_pending_assoc_ind);
 				roam_info->owe_pending_assoc_ind = NULL;
 			} else if (roam_info->ft_pending_assoc_ind) {
+				if (roam_info->ft_pending_assoc_ind->assocReqPtr)
+					qdf_mem_free(roam_info->ft_pending_assoc_ind->assocReqPtr);
 				qdf_mem_free(roam_info->ft_pending_assoc_ind);
 				roam_info->ft_pending_assoc_ind = NULL;
 			}
-- 
2.34.1

